# UTS Data Science - Dokumentasi Praktikum
## Pertemuan 2: Struktur Data Python, NumPy & Pandas

* **Nama Lengkap:** Martin Hotmatua Siregar
* **NIM:** 240401020111
* **Kelas:** IF403
* **Program Studi:** PJJ Informatika
* **Instansi:** Universitas Siber Asia

In [1]:
# Mengimpor library yang dibutuhkan sesuai instruksi modul
import pandas as pd
import numpy as np

# Memuat dataset Titanic langsung dari URL raw GitHub resmi yang valid
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)

# Menampilkan 5 baris pertama untuk memastikan data berhasil dimuat
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


### Q1: Berapa jumlah baris dan kolom pada dataset Titanic? Kolom apa saja yang tersedia?

In [2]:
# Mengecek dimensi dataset (baris, kolom)
baris, kolom = df.shape
print(f"Jumlah Baris: {baris}")
print(f"Jumlah Kolom: {kolom}\n")

# Menampilkan daftar nama kolom yang tersedia
print("Daftar kolom yang tersedia:")
print(df.columns.tolist())

Jumlah Baris: 891
Jumlah Kolom: 12

Daftar kolom yang tersedia:
['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']


**Analisis Temuan Q1:**
Berdasarkan hasil eksekusi kode, dataset Titanic ini memiliki data sebanyak 891 baris penumpang dan terdiri dari 12 kolom informasi. Kolom-kolom tersebut merangkum identitas serta kondisi penumpang, seperti PassengerId, Survived (status keselamatan), Pclass (kelas penumpang), Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, dan Embarked.

### Q2: Kolom mana saja yang memiliki missing values? Berapa jumlah dan proporsinya (%) per kolom?

In [3]:
# Menghitung jumlah missing values per kolom
missing_count = df.isnull().sum()

# Menghitung proporsi persentase missing values
missing_percentage = df.isnull().mean() * 100

# Menggabungkan hasil ke dalam sebuah DataFrame agar rapi
missing_table = pd.DataFrame({
    'Jumlah Missing': missing_count,
    'Proporsi (%)': missing_percentage
})

# Menampilkan kolom yang hanya memiliki nilai kosong (lebih besar dari 0)
print(missing_table[missing_table['Jumlah Missing'] > 0].round(2))

          Jumlah Missing  Proporsi (%)
Age                  177         19.87
Cabin                687         77.10
Embarked               2          0.22


**Analisis Temuan Q2:**
Ternyata ada 3 kolom di dalam dataset ini yang datanya bolong atau tidak lengkap:
1. Kolom **Age (Usia)** kehilangan 177 data (sekitar 19.87%). Nilai kosong ini nanti perlu diantisipasi jika ingin melakukan analisis mendalam tentang usia.
2. Kolom **Cabin (Nomor Kabin)** memiliki tingkat kerusakan data paling parah, yaitu kehilangan 687 data atau sekitar 77.1%. Karena mayoritas datanya kosong, kolom ini kurang ideal untuk analisis statistik langsung.
3. Kolom **Embarked (Pelabuhan Asal)** hanya kehilangan 2 data saja (0.22%), sehingga kondisinya masih sangat aman.

### Q3: Berapa persen penumpang yang selamat (Survived=1)?

In [4]:
# Menghitung persentase keselamatan menggunakan rata-rata kolom Survived dikali 100
persen_selamat = df['Survived'].mean() * 100
print(f"Persentase penumpang yang selamat: {persen_selamat:.2f}%")

Persentase penumpang yang selamat: 38.38%


**Analisis Temuan Q3:**
Dari total keseluruhan penumpang yang tercatat di dataset ini, hanya ada sekitar 38.38% penumpang yang berhasil selamat dari tragedi tenggelamnya kapal Titanic. Sisanya, yaitu sekitar 61.62% penumpang, dinyatakan tidak selamat.

### Q4: Filter penumpang wanita dari kelas 1 (Pclass=1). Berapa jumlahnya dan berapa rata-rata usianya?

In [5]:
# Memfilter data menggunakan boolean indexing untuk wanita di kelas 1
wanita_kelas1 = df[(df['Sex'] == 'female') & (df['Pclass'] == 1)]

# Menghitung jumlah dan rata-rata usianya
jumlah_wanita_k1 = len(wanita_kelas1)
rata_usia_wanita_k1 = wanita_kelas1['Age'].mean()

print(f"Jumlah penumpang wanita kelas 1: {jumlah_wanita_k1} orang")
print(f"Rata-rata usia mereka: {rata_usia_wanita_k1:.1f} tahun")

Jumlah penumpang wanita kelas 1: 94 orang
Rata-rata usia mereka: 34.6 tahun


**Analisis Temuan Q4:**
Ketika kita menyaring data khusus untuk kelompok penumpang wanita yang berada di kelas 1, kita menemukan ada total 94 orang. Kelompok wanita kelas elite ini memiliki rata-rata usia sekitar 34.6 tahun.

### Q5: Gunakan groupby untuk menghitung tingkat keselamatan (mean Survived) per kelas penumpang. Kelas mana yang paling tinggi?

In [6]:
# Mengelompokkan data berdasarkan Pclass dan menghitung rata-rata keselamatan
tingkat_selamat_per_kelas = df.groupby('Pclass')['Survived'].mean() * 100
print(tingkat_selamat_per_kelas.round(2))

Pclass
1    62.96
2    47.28
3    24.24
Name: Survived, dtype: float64


**Analisis Temuan Q5:**
Melalui pengelompokkan data berdasarkan kelas tiket (Pclass), terlihat kontras perbedaan peluang keselamatan yang sangat jelas:
* Kelas 1 memiliki tingkat keselamatan tertinggi mencapai **62.96%**.
* Kelas 2 berada di angka **47.28%**.
* Kelas 3 memiliki tingkat keselamatan paling rendah, yaitu hanya **24.24%**.

Temuan ini membuktikan adanya prioritas penyelamatan bagi penumpang kelas atas saat evakuasi darurat berlangsung di kapal.

### Q6: Analisis Tambahan: Apakah tarif tiket (Fare) yang dibayar penumpang berpengaruh terhadap peluang keselamatan mereka?

In [7]:
# Membandingkan rata-rata tarif tiket (Fare) antara kelompok yang selamat dan tidak selamat
rata_tarif_keselamatan = df.groupby('Survived')['Fare'].mean()

print("Rata-rata tarif tiket berdasarkan status keselamatan:")
print(f"Tidak Selamat (0) : ${rata_tarif_keselamatan[0]:.2f}")
print(f"Selamat (1)       : ${rata_tarif_keselamatan[1]:.2f}")

Rata-rata tarif tiket berdasarkan status keselamatan:
Tidak Selamat (0) : $22.12
Selamat (1)       : $48.40


**Analisis Temuan Q6:**
Sebagai analisis tambahan, saya membedah korelasi antara tarif tiket (Fare) dengan keselamatan. Hasilnya menunjukkan tren yang sangat menarik. Penumpang yang tidak selamat rata-rata hanya membayar tiket seharga $22.12. Sementara itu, penumpang yang selamat rata-rata membeli tiket seharga $48.40 (dua kali lipat lebih mahal). Hal ini semakin memperkuat kesimpulan sebelumnya bahwa faktor ekonomi atau kelas akomodasi sangat berpengaruh besar pada peluang bertahan hidup di tragedi Titanic.

## Kesimpulan & Refleksi Pembelajaran (Sesi 2)

### 1. Apa yang Dipelajari?
Pada pertemuan kedua ini, saya memperdalam kemampuan manipulasi data tabular menggunakan kombinasi library NumPy dan Pandas. Saya belajar cara membaca file data dari URL internet secara langsung, melakukan pengkondisian baris (*filtering* data ganda) menggunakan operator logika, serta mengelompokkan data menggunakan fungsi `groupby` dan agregasi statistik deskriptif dasar.

### 2. Temuan Utama
* Melalui eksplorasi data Titanic, ditemukan bahwa tingkat keselamatan penumpang sangat dipengaruhi oleh kelas sosial (Pclass) dan tarif tiket (Fare). Penumpang kelas 1 memiliki peluang selamat jauh lebih tinggi (62.96%) dibandingkan kelas 3 (24.24%).
* Kualitas data riil di industri seringkali tidak sempurna. Terbukti dari adanya data kosong (*missing values*) yang cukup besar pada variabel krusial seperti usia penumpang (19.87%) dan akomodasi nomor kabin (77.1%).

### 3. Keterbatasan & Pertanyaan yang Muncul
* **Keterbatasan:** Pada analisis ini, saya belum melakukan penanganan apa pun terhadap data yang bolong (*missing values*), seperti mengisi baris yang kosong atau menghapusnya. Data dibiarkan apa adanya sehingga perhitungan statistik usia bisa jadi sedikit bias.
* **Pertanyaan:** Bagaimana taktik terbaik untuk mengisi data kosong pada kolom numerik seperti usia agar tidak merusak distribusi asli datanya? Serta bagaimana cara efektif membuang kolom yang kerusakannya terlalu parah seperti nomor kabin?